db/transaction.go

```diff
@@ -0,0 +1,157 @@
+package db
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+
+	"github.com/attic-labs/noms/go/types"
+
+	"roci.dev/diff-server/kv"
+	nomsjson "roci.dev/diff-server/util/noms/json"
+	"roci.dev/diff-server/util/time"
+)
+
+var (
+	// ErrClosed is the error returned from operations on a Transaction when
+	// it has already been closed.
+	ErrClosed = errors.New("Transaction is closed")
+)
+
+type Transaction struct {
```

Comment:
I think you should make a statement here about the threadsafety of Transaction and I *think* we need to ensure that it is threadsafe. (Even if we have a mechanism in bindings that serializes access, it just seems safer). I think this means:
- guard closed with a mutex
- guard mapeditor operations put/get/has/del with a mutex. (mapeditor is not safe for concurrent access https://github.com/attic-labs/noms/blob/39057233bfdde4b20dbfc8fa8c08a45f0fbe9360/go/types/map_editor.go#L19 -- no locks around access to mapeditor.edits)

cc @aboodman 
-

Reply:
I added a mutex around the operations.

I'll update the comment too.

Reply:
@phritz thanks for thinking about this, you are correct. This is something we need to agree on the design around and document in the class comments of `Transaction` and `db` carefully.

I think it would be reasonable to decide that all access to `repm` is (*must be*) serialized externally, or it would be reasonable to decide that it can be accessed concurrently.

Given the design of Noms and the transaction system we have proposed on top of it in the bug associated with this PR, it **seems** reasonable to me to allow concurrent acccess to `db` and `Transaction`. And if that's what we're doing, @phritz feedbacks are correct.

If either of you have strong reasons why access to db *must be* allowed to be concurrent or *must not be* allowed to be, that would be useful to document.

Reply:
i think if we dont have concurrent access to db then all db activity has to stop during sync, which doesn't sound good. no concurrent access to db also has a fishy smell: databases are not usually serial so i never considered that that was the case. my guess is there is there'd be more undesirable side effects if it were. dunno if those are strong reasons.

also i think(?) the existence of the transaction machinery at the level of repm/client implies to me that concurrent access to Transaction is possible. otherwise i think(?) if access is strictly serialized at the bindings layer then that eliminates the need for a lot of the transaction machinery at the repm/client layer.

disclaimer: i'm new